src/System.Linq.Expressions/src/System/Linq/Expressions/StackGuard.cs

```diff
@@ -0,0 +1,96 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Linq.Expressions
+{
+    internal sealed class StackGuard
+    {
+        private const int MaxExecutionStackCount = 1024;
+
+        private int _executionStackCount;
+
+        public bool TryEnterOnCurrentStack()
```

Comment:
Removed counting of recursion depth and the `Exit` method. A micro-benchmark shows that `EnsureEfficientExecutionStack` executes very fast and the additional cost counting and checking makes it a tiny bit worse (see comments in previous iteration for the numbers). I've kept this change in a separate commit so can revert if there are any strong opinions to keep it (@VSadov, @stephentoub). I'll rebase/squash all commits once we've agreed on the final shape.